Phase 4 of XLA int8 convolution on CUDA

yongfeng-nv opened the pull request:
This is a breakdown of previous PR (https://github.com/tensorflow/tensorflow/pull/29158) per @timshen91 's suggestion.
PLEASE NOTE: This PR DOES depend on (https://github.com/tensorflow/tensorflow/pull/30761, https://github.com/tensorflow/tensorflow/pull/30762,  and https://github.com/tensorflow/tensorflow/pull/30771).

1. Change cudnn_pad_for_tensor_cores to cudnn_pad_for_convolutions and add padding for integer convolution to it.
2. Remove checks for same types in convolution shape inference to allow mixing int8 activation with float bias.
3. Rename InitializeFloatBuffer to InitializeBuffer and add S8 support to it.
4. Add and modify tests for cudnn_pad_for_convolutions_test, cudnn_conv_rewriter_test.cc, and cudnn_fused_conv_rewriter_test.cc.

yongfeng-nv committed:
Phase 4 of XLA int8 convolution on CUDA
1. Change cudnn_pad_for_tensor_cores to cudnn_pad_for_convolutions and add padding for integer convolution to it.
2. Remove checks for same types in convolution shape inference to allow mixing int8 activation with float bias.
3. Rename InitializeFloatBuffer to InitializeBuffer and add S8 support to it.
4. Add and modify tests for cudnn_pad_for_convolutions_test, cudnn_conv_rewriter_test.cc, and cudnn_fused_conv_rewriter_test.cc.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h:
```diff
@@ -0,0 +1,47 @@
+// add slice instruction to remove unnecessary output features.
+class CudnnPadForConvolutions : public HloModulePass {
+ public:
+  explicit CudnnPadForConvolutions(bool is_volta_or_later)
```
One review comment from the previous PR is to let CudnnPadForConvolutions take StreamExecutor*. 
It makes HLO unit test difficult.  I am not sure whether it is possible for a HLO unit test to pass a StreamExecutor to construct a CudnnPadForConvolutions object.  Please see a use case in cudnn_pad_for_convolutions_test.cc @ line 46.

timshen91 replied:
Regarding your comment in the old PR, take a look at gemm_rewrite_test.cc and `default_executor()`. Then it should be easy for `Run()` to take a `StreamExecutor*`.

yongfeng-nv replied:
The default_executor has both major and minor versions set to -1.  I don't know how to set versions to pretend Volta.  The tests copied from cudnn_conv_pad_for_tensor_cores_test however are only for Volta.

timshen91 replied:
Hmm let's not dig into this too deeply, as it's not a big deal. Keeping the bool is good.

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc:
```diff
@@ -631,6 +642,46 @@ TEST_F(CudnnConvRewriterTest, BackwardInputConvolveConstantFilter) {
+    })");
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+
+  //   ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ValueOrDie());
```
I think it's sufficient to `ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ok())`.

yongfeng-nv replied:
I tried the recommended change.  Although I could see the expected error message in the test log, the test didn't finish properly.

[ RUN      ] CudnnConvRewriterTest.ForwardIntegerOutputConvolution
2019-07-18 05:59:31.063159: I tensorflow/compiler/xla/service/platform_util.cc:78] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2019-07-18 05:59:31.063200: I tensorflow/compiler/xla/service/platform_util.cc:78] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2019-07-18 05:59:31.063232: I tensorflow/compiler/xla/service/platform_util.cc:78] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2019-07-18 05:59:31.063246: I tensorflow/compiler/xla/service/platform_util.cc:78] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2019-07-18 05:59:31.068303: F tensorflow/stream_executor/lib/statusor.cc:34] Attempting to fetch value instead of handling error Unimplemented: The convolution instruction with integer inputs only allows float outputs.  Insert a clamp instruction with range [-128, 127) followed by a convert instruction after the convolution instruction for int8 outputs.
external/bazel_tools/tools/test/test-setup.sh: line 310:   549 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1

timshen91 replied:
Have you tried linking `//tensorflow/compiler/jit:xla_gpu_jit` to the test target?

yongfeng-nv replied:
What do you mean linking //tensorflow/compiler/jit:xla_gpu_jit to the test target?  How to do that?  The test doesn't specify dependency on xla_gpu_jit.

Is it because I use ().ValueOrDie() instead of passing status back to the call stack?

timshen91 replied:
XLA defines several Bazel targets as "plugins". They are cc_libraries that the final binary (e.g. a test target) may choose to depend on or not. For GPU plugin, if the test doesn't depend on it, all GPU related code is not linked into the final binary. The error message is saying sort of that.

I think the word "linking" is confusing, but it means to add //tensorflow/compiler/jit:xla_gpu_jit to the test target as a dependency.

yongfeng-nv replied:
Yes, "linking //tensorflow/compiler/jit:xla_gpu_jit" makes the test working.  I have made another commit for the changes.

I saw the message about "linking" in the log and considered it information message.  Is GPU plugin required by error handling, but not by regular HLO pattern matching tests.

yongfeng-nv committed:
1. Add //tensorflow/compiler/jit:xla_gpu_jit to cudnn_conv_rewriter_test's dependency.
2. Enable error checking CudnnConvRewriter.ForwardIntegerOutputConvolution.

gbaned commented:
@yongfeng-nv Could you please resolve the conflicts? Thanks!

yongfeng-nv committed:
Mapping integer convolutions to CuDNN with pattern matching.
1. New and updated tests.
2. Revert changes in shape_inference.cc.
3. Other minor changes to enable int8 convolution and for formatting.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/buffer_comparator.cc:
```diff
@@ -405,11 +405,13 @@ StatusOr<bool> HostCompare(se::Stream* stream, se::DeviceMemoryBase lhs,
         return a;
       }
-      return std::max(-kMaxFp16Value, std::min(a, kMaxFp16Value));
+      return std::max(static_cast<ComparisonType>(-kMaxFp16Value),
```
This change is needed, because 65505 can't fit in int8.  Refactoring this lambda to a set of template/overloaded helper functions is another solution.

yongfeng-nv commented on tensorflow/compiler/xla/service/shape_inference.cc:
```diff
@@ -1598,6 +1598,11 @@ ShapeInference::InferDegenerateDimensionBroadcastShape(HloOpcode operation,
         batch_group_count, feature_group_count);
   }
 
+  if (!ShapeUtil::SameElementTypeIgnoringFpPrecision(lhs, rhs)) {
```
Revert this change from the previous commit.

yongfeng-nv committed:
Revert one line of unnecessary change.

yongfeng-nv committed:
1. Add/modify tests for int8 convolution pattern matching implementation.
2. Add prelimiary support for integer types to InitializeTypedBuffer.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:
```diff
@@ -40,30 +41,16 @@ class CudnnFusedConvRewriterTest : public HloTestBase {
-                       {"ALPHA_CONV_SCALAR", alpha_conv_scalar},
-                       {"ALPHA_SIDE_INPUT_SCALAR", alpha_side_input_scalar}});
-      string optimized_hlo_string = GetOptimizedHlo(hlo_resolved_string);
+    for (absl::string_view type : {"f16", "f32", "f64"}) {
```
Revert the changes on floating-point tests, as it is very difficult to use them to test int8 convolutions.  We created more int8 specific tests.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:
```diff
@@ -432,6 +441,207 @@ TEST_F(CudnnFusedConvRewriterTest, TestFusedConvClamp) {
+    })");
+}
+
+TEST_F(CudnnFusedConvRewriterTest, TestFusedConvWithScaledInt8SideInputBiasInt8ToFloat) {
```
No need to have extra check for convert<float>(int8_side_input) in the result HLO, because if it is absorbed by mistake, CuDNN won't run properly and fail the test.

gbaned commented:
@yongfeng-nv Could you please resolve the conflicts? Thanks!

mellanox-github commented:
Can one of the admins verify this patch?

cheshire commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_algorithm_picker.cc:
```diff
@@ -271,10 +271,10 @@ StatusOr<AutotuneResult> CudnnConvAlgorithmPicker::PickBestAlgorithmNoCache(
+  const auto initialize_buffer = [&stream, &rng_state](
+                                     DeviceMemoryBase buffer,
+                                     const Shape& buffer_shape) {
+    InitializeBuffer(stream, buffer_shape.element_type(), &rng_state, buffer);
```
What changed here? Isn't buffer_shape.element_type() the same as result_shape.element_type() ?

yongfeng-nv replied:
initialize_buffer has two call sites down here.  Before this change, both took result_shape.  Now, since one takes operand->shape() and the other takes result_shape, we need to move buffer_shape from from captures to arguments for the lambda.

gbaned commented:
@yongfeng-nv Could you please resolve the conflicts? Thanks!

gbaned commented:
@yongfeng-nv  Still, conflicts appearing. Could you please resolve those? Thanks!

yongfeng-nv commented:
Close this PR, as its change has been moved to https://github.com/tensorflow/tensorflow/pull/30771 and https://github.com/tensorflow/tensorflow/pull/31988.
